Milestone 2 [Release Automation]: Updates to release_info for automatic changelog creation

ankita240796 opened the pull request:
## Explanation

Updates `release_info.py`:

- Accept personal_access_token through interactive prompt
- Check for blocking bugs and pull requests in current release before changelog creation
- Automatic generation of changelog

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

ankita240796 committed:
Get personal access token through prompt

ankita240796 committed:
Add check for blocking bugs

ankita240796 committed:
Add test for PRs for current release

ankita240796 committed:
Automate updates to changelog and authors

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
     return [item for item in diff_list if item.startswith('core/storage')]
 
 
+def check_blocking_bugs(repo):
```
will it be possible to do this without passing the whole object?

Should we create a `GithubRepo` class which will handel all this method?

ankita240796 replied:
Yes, we can do that. We can also use a global variable instead. Which do you suggest?

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
+    Returns:
+        int: Number of unresolved blocking bugs.
+    """
+    blocking_bugs_milestone = repo.get_milestone(number=39)
```
Should we move this number and other such constants to feconf or on top? (not sure whether we should use feconf for release script)

@seanlip, do you have any suggestion on this?

ankita240796 replied:
I think it is fine to add it on top with a comment.

seanlip replied:
In general, it is fine to keep constants at the top of a script if the script is the only file which uses them.

So, in this case, yep, please add it on top with a comment.

ankita240796 replied:
Done.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
+
+def check_prs_for_current_release(repo):
+    """Checks that all pull requests for current release have a
+    PR: released label.
```
Should we add a quote (') to highlight the label name?

ankita240796 replied:
Done.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
+        int. Whether all pull requests for current release have a
+            PR: released label.
+    """
+    all_prs = repo.get_pulls(state='all')
```
will it fetch all the PRs? (If this will be expensive then we can create a separate milestone for release related PRs.)

ankita240796 replied:
Yes, it fetches all the PRs since there is no method in pygithub to filter the PRs. I tried this and it was not time-expensive. It works fine. You can try it once to check if it works fine and let me know. I will make the updates accordingly.

ankita240796 replied:
On further thoughts, I feel that this may become expensive as the number of PRs keeps increasing. I am not sure if adding a milestone for release related PRs will help us to tackle this because we can only obtain the number of open/closed PRs in the milestone not the details of the PR. So, should we instead make a get request instead and fetch PRs with both the labels and compare the output?

seanlip replied:
If you can only fetch the PRs with the "for release" label, that might be better. Then you can iterate along them and ensure that they all have the "released" label as well.

But if that is not feasible, then fetch all PRs for safety. If you add a milestone then it becomes confusing, since we use milestones to track tasks (issues) and not PRs. Plus, this is susceptible to human error.

Also, I'm a bit confused about one thing -- you mentioned that there's no way in pygithub to filter the PRs, but you also suggested making a get request to fetch PRs with both the labels (which seems like a sort of filtering). So ... what is the exact situation here?

ankita240796 replied:
I did not find any function in pygithub to fetch pull requests using a filter but the solution on stackoverflow [here](https://stackoverflow.com/a/36984550) provides a way to fetch using a get request through github api.

seanlip replied:
Ah ok, thanks for clarifying.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
+        repo: github.Repository.Repository. The PyGithub object for the repo.
+
+    Returns:
+        int. Whether all pull requests for current release have a
```
bool*

ankita240796 replied:
Done.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
     return [item for item in diff_list if item.startswith('core/storage')]
 
 
+def check_blocking_bugs(repo):
```
Maybe `get_blocking_bug_issue_count`?

ankita240796 replied:
Done.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
+    return blocking_bugs_milestone.open_issues
+
+
+def check_prs_for_current_release(repo):
```
Maybe `check_prs_for_current_release_are_released` or `check_blocking_bug_fix_released`? (The current name doesn't explains what the method checks.)

ankita240796 replied:
Done.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
+        filepath: str. The path of the file to update.
+        new_list: list(str). The list of new authors or contributors to
+            add to the file.
+        start_index_line: str. The line whose index determines the start
```
int or str?

ankita240796 replied:
str - This is the line for which the function calculates the index.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
-    if parsed_args.personal_access_token is None:
+    personal_access_token = getpass.getpass(
+        prompt=(
+            'Personal access token for your github ID. '
```
If this is going to be a message for asking the user to provide the PAT then let's add `Please provide your GitHub ....` to make it clear for the user.

ankita240796 replied:
Done.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
-            'to the script using --personal_access_token=<token>')
+            'access token at https://github.com/settings/tokens and re-run '
+            'the script')
+        return
```
Let's add line 389 over here so that all the `personal_access_token` related verification are done at once. (Currently, the username verification is in between.)

ankita240796 replied:
Done.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
+        return
+
+    parsed_args = _PARSER.parse_args()
+    if parsed_args.github_username is None:
```
If this is a mandatory requirement then let's move it on top*

ankita240796 replied:
Done.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
+            'is%3Aissue+milestone%3A%22Blocking+bugs%22')
+        raise Exception(
+            'There are %s unresolved blocking bugs. Please '
+            'ensure that they are resolved before changelog creation.' % (
```
Should this message state that we have opened the browser with a link?

ankita240796 replied:
We don't do that at other place as well. Should I update the function in `common.py` to always print this message?

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -292,24 +303,111 @@ def _check_storage_models(current_release):
+        raise Exception(
+            'There are PRs for current release which do not have '
+            'a \'PR: released\' label. Please ensure that they are released '
+            'before changelog creation.')
```
Ditto*

ankita240796 replied:
Replied.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -391,7 +493,7 @@ def main():
             'possible.``\n' % existing_author_comma_list)
 
-        if parsed_args.personal_access_token:
+        if personal_access_token:
```
will this script work without personal_access_token?

ankita240796 replied:
No.

ankita240796 replied:
No, we need the token to obtain the repo.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -411,6 +513,156 @@ def main():
+    # the script again.
+    try:
+        with python_utils.open_file(
+            '../release_summary.md', 'r') as release_summary_file:
```
if this filename is used in multiple place then let's create a constant var.

ankita240796 replied:
Done.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -411,6 +513,156 @@ def main():
+                    changelog_file.write(line)
+            python_utils.PRINT('Updated Changelog!')
+
+            python_utils.PRINT('Updating Authors...')
```
Should we create methods like `update_changelog`, `update_author` etc. and make the main method more readable? (I'm fine with this structure as well.)

ankita240796 replied:
I created a function to do the common handling but I feel it will be better to have separate functions for readability. I will make the updates.

ankita240796 replied:
Done.

DubeySandeep commented on scripts/release_info.py:
```diff
@@ -411,6 +513,156 @@ def main():
+                        nxt_line = (
+                            '            <p translate="I18N_ABOUT_PAGE_CREDITS_'
+                            'TAB_TEXT_BOTTOM" translate-values="{listOfNames: '
+                            '$ctrl.listOfNames}">\n')
```
Should we add a comment around this line of about page html file? 

Maybe we can create `contributors-list` directive which will only contain the list of names? (not sure about this)

ankita240796 replied:
Added a comment.

I am sorry, I did not get the second part clearly. Are you suggesting to have a separate directive to only display the credits?

DubeySandeep reviewed (commented):
@ankita240796, Thanks for the PR, I've left a few comments PTAL!

ankita240796 committed:
Address review comments

ankita240796 commented:
Hi @DubeySandeep, I have addressed all the review comments, PTAL. Thanks!

ankita240796 committed:
Merge develop

seanlip commented on core/templates/dev/head/pages/about-page/about-page.directive.html:
```diff
@@ -102,6 +102,7 @@ <h2 translate="I18N_ABOUT_PAGE_CREDITS_TAB_HEADING"></h2>
             <p translate="I18N_ABOUT_PAGE_CREDITS_TAB_TEXT">
             </p>
             <div class="oppia-about-credits-letter-groups three-col">
+              <!-- The indentation used here should match the value of SPAN_INDENT and LI_INDENT variables defined  in release_info.py script-->
```
Drop one of the spaces before "in".

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -37,6 +41,7 @@
 # pylint: enable=wrong-import-position
 
 GIT_CMD_GET_STATUS = 'git status'
+GIT_CMD_CHECKOUT = 'git checkout -- %s'
```
For any string that still needs interpolation, start its name with GIT_CMD_TEMPLATE_

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -51,16 +56,29 @@
                     'CURRENT_COLLECTION_SCHEMA_VERSION']
 FIRST_OPPIA_COMMIT = '6a7138f5f603375e58d1dc3e1c4f1c80a126e249'
 NO_LABEL_CHANGELOG_CATEGORY = 'Uncategorized'
+# This should match the indentation of span and li elements used
```
I think you could have the script test/assert this? Better than just leaving a comment.

Could do it in a backend test (scripts/release_info_test.py is fine) so that if someone changes the directive and indents everything 2 spaces to the right (say), then that will be caught and these constants will be fixed.

ankita240796 replied:
Instead of adding a test, should I do this: select the line which contains `<span>A</span>` and then find the indentation using this line (same for li element) since a test will also need to select this line.

seanlip replied:
Yes, that's fine too. And throw an error if there's no such line. (In general be careful about handling all the possible error cases and being noisy about them, since there's no guarantee that the files you're working with won't change in the future.)

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -51,16 +56,29 @@
+# core/templates/dev/head/pages/about-page/about-page.directive.html.
+SPAN_INDENT = '              '
+LI_INDENT = '                '
+# This line should match the line after div element for credits in
```
Same thing, try and test this to ensure that no incoming PR can break this script's functionality.

ankita240796 replied:
Replied above.

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -51,16 +56,29 @@
+LINE_AFTER_Z_CREDITS = (
+    '            <p translate="I18N_ABOUT_PAGE_CREDITS_'
+    'TAB_TEXT_BOTTOM" translate-values="{listOfNames: $ctrl.listOfNames}">\n')
+ABOUT_PAGE_FILEPATH = (
```
Should generally use os.path.join for platform-independence.

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -292,6 +310,188 @@ def _check_storage_models(current_release):
+        filepath: str. The path of the file to update.
+        new_list: list(str). The list of new authors or contributors to
+            add to the file.
+        start_index_line: str. The line whose index determines the start
```
line --> content of the line

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -292,6 +310,188 @@ def _check_storage_models(current_release):
+    content = []
+    with python_utils.open_file(filepath, 'r') as f:
+        content = f.readlines()
+    start_index = content.index(start_index_line) + 2
```
In a comment, explain the "+ 2".

It also seems to suggest that the start_index_line is 2 lines before the actual start, in which case a renaming might be appropriate.

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -292,6 +310,188 @@ def _check_storage_models(current_release):
+        start_index_line: str. The line whose index determines the start
+            of authors/contributors list in the file.
+    """
+    content = []
```
file_lines or similar, perhaps. "content" is a bit vague, sounds like it could be a string.

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -292,6 +310,188 @@ def _check_storage_models(current_release):
+
+    Args:
+        filepath: str. The path of the file to update.
+        new_list: list(str). The list of new authors or contributors to
```
Should probably add checks to ensure this doesn't intersect with any existing lines in the file?

ankita240796 replied:
Should we make it into a `list(set(list` instead?

seanlip commented on scripts/release_info.py:
```diff
@@ -292,6 +310,188 @@ def _check_storage_models(current_release):
+    current_date = datetime.date.today().strftime('%d %b %Y')
+    python_utils.PRINT('Updating Changelog...')
+    start_index = release_summary_content.index('### Changelog:\n') + 1
+    end_index = release_summary_content.index('### Commit History:\n')
```
This seems dependent on release_summary_content not having a section between Changelog and Commit History, which may not be preserved in the future. You might want to implement a backend test that asserts this condition is always satisfied.

ankita240796 replied:
Added a test in the script itself.

seanlip commented on scripts/release_info.py:
```diff
@@ -21,13 +21,17 @@
 
 import argparse
```
Try to find a way to add unit tests for as many functions in this file as possible.

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -292,6 +310,188 @@ def _check_storage_models(current_release):
+        release_summary_content: list(str). List of lines in
+            ../release_summary.md.
+    """
+    python_utils.PRINT('Updating Authors...')
```
Here and below: Authors --> AUTHORS file.

Ditto for CONTRIBUTORS.

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -292,6 +310,188 @@ def _check_storage_models(current_release):
+    python_utils.PRINT('Updating Authors...')
+    start_index = release_summary_content.index(
+        '### New Authors:\n') + 1
+    end_index = release_summary_content.index(
```
Same note about adding tests to ensure these sections are adjacent in the release summary file. In those backend tests, add a comment about why the check is placed there (it is so that this script can run).

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -292,6 +310,188 @@ def _check_storage_models(current_release):
+    python_utils.PRINT('Updating Contributors...')
+    start_index = release_summary_content.index(
+        '### New Contributors:\n') + 1
+    end_index = release_summary_content.index(
```
Ditto.

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -292,6 +310,188 @@ def _check_storage_models(current_release):
+        contributor.split('<')[0] for contributor in new_contributors]
+    new_credits.sort()
+    credit_dict = collections.defaultdict(list)
+    for credit in new_credits:
```
credit --> developer_name

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -292,6 +310,188 @@ def _check_storage_models(current_release):
+    return True
+
+
+def update_sorted_file(filepath, new_list, start_index_line):
```
Actually, I just realized that a bunch of these functions are in the wrong place. release_info.py is a script that generates a doc, release_info.md. It should not be changing files or have other side effects -- that should happen as part of the release process in a separate script that actually does the release process.

I think people expect release_info to be like a GET action, i.e. it generates an info doc outside the oppia folder and doesn't modify state elsewhere. We should preserve that.

ankita240796 replied:
Updated.

seanlip commented on scripts/release_info.py:
```diff
@@ -409,7 +642,83 @@ def main():
+        python_utils.PRINT(
+            'Please update %s to:\n'
+            '- have a correct changelog for '
+            'updating the CHANGELOG file\n'
```
You can probably move some of this to the previous line?

ankita240796 replied:
Done.

seanlip commented on scripts/release_info.py:
```diff
@@ -409,7 +642,83 @@ def main():
+    source_branch = 'develop'
+    target_branch = 'update-changelog-for-releasev%s' % current_release_version
+    release_summary_content = []
+    # This complete code block is wrapped in try except since in case of
```
Good call!

ankita240796 replied:
Thank you!

seanlip reviewed (changes requested):
Thanks @ankita240796, took a pass!

kevinlee12 commented on scripts/release_info.py:
```diff
@@ -51,16 +56,29 @@
-        'Personal access token for your github ID. '
-        'You can create one at https://github.com/settings/tokens'),
-    type=str)
+    '--github_username', help=('Your GitHub username. '), type=str)
```
Remove space after the period.

ankita240796 replied:
Done.

kevinlee12 reviewed (commented):
other than Sean's and Sandeep's comments, I have one more minor one.

ankita240796 committed:
Address review comments

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into auto-changelog

ankita240796 committed:
Replace content with lines

ankita240796 committed:
Separate files

seanlip commented on core/templates/dev/head/pages/about-page/about-page.directive.html:
```diff
@@ -102,6 +102,7 @@ <h2 translate="I18N_ABOUT_PAGE_CREDITS_TAB_HEADING"></h2>
             <p translate="I18N_ABOUT_PAGE_CREDITS_TAB_TEXT">
             </p>
             <div class="oppia-about-credits-letter-groups three-col">
+              <!-- The indentation used here should match the value of SPAN_INDENT and LI_INDENT variables defined in release_info.py script-->
```
If you're deriving this proactively based on the indentation level of `<span>A</span>` (as I think you suggested in a reply to a review comment), then you don't need this comment.

ankita240796 replied:
Updated.

seanlip commented on scripts/common.py:
```diff
@@ -23,6 +23,19 @@
+    Raises subprocess.CalledProcessError upon failure.
+
+    Args:
+        cmd_str: str. The command string to execute
```
Add period at end.

ankita240796 replied:
Done.

seanlip commented on scripts/common.py:
```diff
@@ -23,6 +23,19 @@
+    Returns:
+        str. The output of the command.
+    """
+    return subprocess.check_output(cmd_str.split(' ')).strip()
```
I don't think you need this function, actually. Just pass in a list of "command tokens" from the get-go every time you call subprocess.check_output; that's pretty standard.

ankita240796 replied:
Done.

ankita240796 replied:
Just to confirm - you meant removing the splitting part by passing tokens as a list or removing the function overall? I have made changes only by removing the splitting part. If you meant removing the function, I will make changes according to that.

seanlip replied:
Removing the function overall.

seanlip commented on scripts/create_changelog.py:
```diff
@@ -0,0 +1,329 @@
+# core/templates/dev/head/pages/about-page/about-page.directive.html.
+SPAN_INDENT = '              '
+LI_INDENT = '                '
+# This line should match the line after div element for developer_names in
```
Try to derive most of these directly from the file if possible (throwing an exception if it doesn't exist).

ankita240796 replied:
Done.

seanlip commented on scripts/create_changelog.py:
```diff
@@ -0,0 +1,329 @@
+        filepath: str. The path of the file to update.
+        new_list: list(str). The list of new authors or contributors to
+            add to the file.
+        last_comment_line: str. The content of the line of last comment in
```
I think you can just make this a constant here and not pass it in, it's the same for both files. Throw an exception if it can't be found.

ankita240796 replied:
Done.

seanlip commented on scripts/create_changelog.py:
```diff
@@ -0,0 +1,329 @@
+                repo_fork.update_file(
+                    contents.path, 'Update %s' % filepath, f.read(),
+                    contents.sha, branch=target_branch)
+        common.run_cmd(GIT_CMD_TEMPLATE_CHECKOUT % ('%s %s %s %s') % (
```
I think you need to add an interactive step where the author can review the diff, make changes if necessary, and then say "yes, this is correct, please go ahead".

ankita240796 replied:
Done.

seanlip commented on scripts/create_changelog.py:
```diff
@@ -0,0 +1,329 @@
+#!/usr/bin/env python
```
Perhaps rename file to: update_changelog_and_credits.py

ankita240796 replied:
Done.

seanlip commented on scripts/create_changelog.py:
```diff
@@ -0,0 +1,329 @@
+            repo_fork.get_git_ref('heads/%s' % target_branch).delete()
+        except Exception:
+            python_utils.PRINT(
+                'Please ensure that %s branch is deleted before re-running '
```
Maybe have a check at the start of the script to ensure that the branch to be created isn't yet present. (Better than relying on the developer to read each message carefully -- helps mitigate human error.)

ankita240796 replied:
Done.

ankita240796 commented:
Hi @seanlip, I am still addressing the review comments. I am just replying to the ones that I have completed to keep a track, so please avoid taking a pass now. I will assign you once I am done. Sorry for the trouble. Thanks!

ankita240796 committed:
Rename file, move check to start

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into auto-changelog

ankita240796 committed:
Clean up main

ankita240796 committed:
Check ordering of sections

ankita240796 committed:
Update indent calculation

ankita240796 committed:
Remove extra script, add milestone constant

ankita240796 committed:
Remove extra last comment line

ankita240796 committed:
Fix year

ankita240796 committed:
Add tests

ankita240796 commented:
Hi @seanlip @DubeySandeep @kevinlee12 @nithusha21, I have addressed all the review comments, PTAL. Thanks!

seanlip commented on scripts/common.py:
```diff
@@ -24,6 +24,19 @@
 RELEASE_BRANCH_NAME_PREFIX = 'release-'
 
 
+def run_cmd(cmd_tokens):
```
Actually, looking at this -- I'm fine with keeping it as-is, too; it's a nice abstraction. So, no change needed (though I'm happy either way).

ankita240796 replied:
Updated!

ankita240796 replied:
(To the first one - I liked that more :P)

seanlip commented on scripts/update_changelog_and_credits.py:
```diff
@@ -0,0 +1,454 @@
+    with python_utils.open_file(filepath, 'r') as f:
+        file_lines = f.readlines()
+
+
```
Drop extra line.

ankita240796 replied:
Done.

seanlip reviewed (approved):
Thanks @ankita240796! Approving as codeowner for feconf.py. Noticed a couple other things on a cursory look.

ankita240796 committed:
Remove extra line, update run_cmd

kevinlee12 reviewed (approved):
lgtm!

ankita240796 committed:
Fix test, increase coverage

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into auto-changelog

ankita240796 commented on scripts/update_changelog_and_credits.py:
```diff
@@ -0,0 +1,452 @@
+            're-running the script' % target_branch)
+
+
+def create_branch(repo_fork, target_branch, github_username):
```
Adding test for this function would be basically adding mocks for the pygithub functions like `get_branch`, `create_git_ref` etc just to ensure every line is covered but we cannot check if the branch is actually created.

seanlip replied:
We have two choices:

(a) We could disable this using the no-coverage check -- see my other comment for details.
(b) We could test this trivially using mocking.

I think (b) is probably better, because if you do (a), other devs will just follow it. What do you think? It has the benefit that, if these functions become more complex in the future, we have tests for them that can subsequently be augmented. For now, you could just check that the correct methods are called using a CallCounter (see test_utils).

ankita240796 replied:
Yes, I think I agree with you on this. (b) is a better choice here. I will add the tests accordingly.

ankita240796 replied:
Added tests.

ankita240796 commented on scripts/update_changelog_and_credits.py:
```diff
@@ -0,0 +1,452 @@
+
+
+if __name__ == '__main__':
+    main()
```
How should I test this line?

seanlip replied:
See the bottom of scripts/build.py for what we do in this case.

ankita240796 replied:
Updated according to `scripts/build.py`.

ankita240796 reviewed (commented):
Hi @seanlip, I have a doubt regarding addition of tests for the lines I have commented on. If I do not add tests for them, the coverage decreases. So, how should I handle these two cases? 
Thanks!

ankita240796 committed:
Remove extra line

ankita240796 committed:
Update test, increase coverage

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into auto-changelog

ankita240796 committed:
Fix new lines

ankita240796 commented:
Hi @seanlip @DubeySandeep @nithusha21 PTAL, Thanks!

seanlip commented:
Hi @ankita240796 -- coverage checks are still failing?

seanlip commented on scripts/common.py:
```diff
@@ -24,6 +24,19 @@
 RELEASE_BRANCH_NAME_PREFIX = 'release-'
 
 
+def run_cmd(cmd_str):
```
Hang on ... there may have been a miscommunication here. By "keeping it as-is" I meant using a list of command tokens, as opposed to removing the function completely. Sorry for not being clear!

ankita240796 replied:
Yes, I did not remove the function. I had updated this to use a list of tokens but you mentioned later that it was a nice abstraction and I liked the version of passing string instead of list, so I removed the update. Do you want me to make that update here instead?

seanlip replied:
Ah - to be clear, I had meant that the version of the function that uses list as an input is a nice abstraction, so let's use that please. It seems better to think of the args as a sequence: https://stackoverflow.com/a/15109975

ankita240796 replied:
Yes, got it! Updated now.

ankita240796 committed:
Make coverage 100% for release_info script

ankita240796 committed:
Increase coverage for common

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into auto-changelog

ankita240796 committed:
Fix lint

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into auto-changelog

ankita240796 commented:
Hi @seanlip, the coverage checks should pass now. Earlier some tests were missing for `release_info.py` and `common.py`. I have added all tests now, PTAL. Thanks!

seanlip commented:
Ah - to be clear, I was saying that the version of the function that uses
list is a nice abstraction, so let's use that.

ankita240796 committed:
Fix test

ankita240796 commented:
Hi @seanlip @nithusha21 @DubeySandeep, I have fixed all the tests now, PTAL. Thanks!

seanlip reviewed (approved):
Approving as codeowner. Thanks @ankita240796 !
